src/utils/useCollection/index.ts

```diff
@@ -1,52 +1,48 @@
 import { useState, useEffect } from 'react'
-import firestore from '#utils/useFirestore'
+import firestore from '#utils/firestore'
 
-function useCollection<TDocument>(
+interface CollectionState<TDocument> {
+  data: WithID<TDocument>[]
+  loading: boolean
+  error: Error | null
+}
+
+type DocumentData = firebase.default.firestore.DocumentData
+export type WithID<TDocument extends DocumentData> = TDocument & { id: string }
+
+function useCollection<TDocument extends DocumentData>(
   collectionName: string,
-): UseFirestoreData<DocumentWithId<TDocument>[]> {
-  const [collectionData, setCollectionData] = useState<
-    UseFirestoreData<DocumentWithId<TDocument>[]>
+): CollectionState<TDocument> {
+  const [collectionState, setCollectionState] = useState<
+    CollectionState<TDocument>
   >({
-    state: 'loading',
-    data: null,
+    data: [],
+    loading: true,
     error: null,
   })
 
   useEffect(() => {
-    firestore
-      .collection(collectionName)
-      .get()
-      .then((snapshot) => {
-        if (!snapshot.empty) {
-          setCollectionData({
-            state: 'success',
-            data: snapshot.docs.map<DocumentWithId<TDocument>>(
-              (s) =>
-                ({
-                  id: s.id,
-                  ...s.data(),
-                } as DocumentWithId<TDocument>),
-            ),
-            error: null,
-          })
-        } else {
-          setCollectionData({
-            state: 'not-found',
-            data: null,
-            error: null,
-          })
-        }
-      })
-      .catch((error) => {
-        setCollectionData({
-          state: 'error',
-          data: null,
-          error,
-        })
-      })
+    async function getCollection() {
+      try {
+        const collectionSnapshot = await firestore
+          .collection(collectionName)
+          .get()
+        const documents = collectionSnapshot.docs.map(
+          (doc) =>
+            ({
+              id: doc.id,
+              ...doc.data(),
+            } as WithID<TDocument>),
+        )
+        setCollectionState({ data: documents, loading: false, error: null })
+      } catch (error) {
+        setCollectionState({ data: [], loading: false, error })
+      }
+    }
+    getCollection()
```

Comment:
Why change what wasn't broken? :grey_question: 

Also, you have now lost a case of distinguishing a "Not found" and an error - probably didn't _quite_ make sense for the collection hook, but still...

Reply:
always leave things cleaner than you found them, it was very verbose and made a distinction (not-found) that doesn't actually reflect a real state. when accessing firestore with a bogus key, you just receive an empty result, and using a good key in an empty collection gives the same result. either way you can just check the data.length instead of mimicking a state machine

meaning that possible states, when given names, is loading, success, success_empty (or not found, if you will), and error.

Reply:
also, I did not lose it, I can assign `data: null` on error and make the type of data nullable, but that drags a whole load of aggressive type guarding, I've tried it, it looks bad and is annoying to work with. You just need to handle loading -> error -> data which is convention in [apollo client](https://www.apollographql.com/docs/react/get-started/#request-data) [react-query](https://react-query.tanstack.com/guides/queries) [use-http](https://use-http.com/#/?id=basic-usage-auto-managed-state)  [swr](https://swr.vercel.app/) and many many more. So it's not contrived to work assuming this workflow

Reply:
I rather mimic a finite state machine. This only "looks cleaner" in my opinion and that's not a big priority. The event state machine esque is more declarative and explicit of a pattern, so will look busier, i.e. XState configs look busy and messy, but are extremely explicit and declarative, it's kind of the point.

Reply:
we don't get a more explicit or descriptive state either way, the state returned from the firebase is either success with 0-n elements, or error, I don't see the need to go further than a regular `async` function for such a small pool of states (pending, success, error)

Reply:
The tooling we generally use React-Query and XState would disagree with that, so I am all for finding a middle ground here but this came down to your refactoring based on your preference. 

Cleaner or clever refactors aren't always better. I leave this up to @nobrayner it was his original code. 

I would like to see tests for Refactors though.

Reply:
The biggest issue I have here - and it doesn't relate directly to this code anymore - is that there is no longer a distinction between a "not found/no items" and error state. I'll make a new review on that to more focus the discussion